### PR TITLE
fix(web): ws-hook universal error handler

### PR DESCRIPTION
**Pull request by robot9706**
`WS_TYPE_DYO_ERROR` was handled multiple times by showing a toast, pages which use multiple `useWebSocket` hooks show multiple toasts.

Implemented a universal error handler in `WebSocketClient` which shows the toast, pages can still do their error handling logic using `WS_TYPE_DYO_ERROR`.

Added a WebSocket provider to create the socket only once.

### CODE REVIEW

**Comment by polaroi8d** [comment]
Is this solve that the platform shows error messages multiple times? LGTM (informal) but please wait for others with the merge.

**Review by polaroi8d — APPROVED** [review]
Is this solve that the platform shows error messages multiple times? LGTM (informal) but please wait for others with the merge.

**Comment by robot9706** [comment]
> Is this solve that the platform shows error messages multiple times? LGTM (informal) but please wait for others with the merge.

@polaroi8d Yep.

I'd also like @m8vago to check as well :)

File: web/crux-ui/src/hooks/use-websocket.ts [hunk 1]
```diff
@@ -1,12 +1,19 @@
+import { defaultWsErrorHandler } from '@app/errors'
 import { isServerSide } from '@app/utils'
 import { WebSocketClient, WebSocketClientOptions, WebSocketEndpoint } from '@app/websockets/client'
+import useTranslation from 'next-translate/useTranslation'
 import { useEffect, useRef, useState } from 'react'
 
 let client: WebSocketClient
 
 export const useWebSocket = (route: string, options?: WebSocketClientOptions): WebSocketEndpoint => {
+  const { t } = useTranslation()
```
Comment (m8vago) [inline web/crux-ui/src/hooks/use-websocket.ts:10]: as we discussed, we will have to make a provider for the WebSocketClient, so the TFunction only gets obtained once.

File: web/crux-ui/src/websockets/client.ts [hunk 2]
```diff
@@ -135,6 +139,7 @@ export class WebSocketClient {
   private token?: string
   private unsubscribe?: VoidFunction
   private routes: Map<string, WebSocketRoute> = new Map()
+  private errorHandler?: WsErrorHandler = undefined
```
Comment (m8vago) [inline web/crux-ui/src/websockets/client.ts:142]: use `null` please

File: web/crux-ui/src/websockets/client.ts [hunk 3]
```diff
@@ -318,6 +327,10 @@ export class WebSocketClient {
         this.logger.verbose('Receiving message:', message.type)
         this.logger.verbose('Content:', message.payload)
 
+        if (message.type == WS_TYPE_DYO_ERROR && this.errorHandler) {
```
Comment (m8vago) [inline web/crux-ui/src/websockets/client.ts:330]: `===`

**Review by polaroi8d — APPROVED** [review]
Still okey for me.

File: web/crux-ui/src/providers/websocket.tsx [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+interface WebSocketContextInterface {
+  client: WebSocketClient
+}
+
+export const WebSocketContext = React.createContext<WebSocketContextInterface>({ client: null })
+
+export const WebSocketProvider = (props: React.PropsWithChildren<{}>) => {
+  const { t } = useTranslation()
+
+  const [wsClient, _] = useState(() => {
```
Comment (m8vago) [inline web/crux-ui/src/providers/websocket.tsx:16]: useRef instead?

Comment (robot9706) [reply web/crux-ui/src/providers/websocket.tsx:16]: I wanted to use `useRef` but I need a default value initialiser with a function. `useRef` only supports values.

**Review by m8vago — APPROVED** [review]

**Pull request merged**